src/Development/IDE/LSP/CodeAction.hs

```diff
@@ -102,13 +104,49 @@ suggestAction contents Diagnostic{_range=_range@Range{..},..}
 --     Could not find module ‘Data.Cha’
 --     Perhaps you meant Data.Char (from base-4.12.0.0)
     | "Could not find module" `T.isInfixOf` _message
-    , "Perhaps you meant"     `T.isInfixOf` _message
-      = map proposeModule $ nubOrd $ findSuggestedModules _message where
+    , "Perhaps you meant"     `T.isInfixOf` _message = let
       findSuggestedModules = map (head . T.words) . drop 2 . T.lines
       proposeModule mod = ("replace with " <> mod, [TextEdit _range mod])
+      in map proposeModule $ nubOrd $ findSuggestedModules _message
+
+--  ...Development/IDE/LSP/CodeAction.hs:103:9: warning:
+--   * Found hole: _ :: Int -> String
+--   * In the expression: _
+--     In the expression: _ a
+--     In an equation for ‘foo’: foo a = _ a
+--   * Relevant bindings include
+--       a :: Int
+--         (bound at ...Development/IDE/LSP/CodeAction.hs:103:5)
+--       foo :: Int -> String
+--         (bound at ...Development/IDE/LSP/CodeAction.hs:103:1)
+--     Valid hole fits include
+--       foo :: Int -> String
+--         (bound at ...Development/IDE/LSP/CodeAction.hs:103:1)
+--       show :: forall a. Show a => a -> String
+--         with show @Int
+--         (imported from ‘Prelude’ at ...Development/IDE/LSP/CodeAction.hs:7:8-37
+--          (and originally defined in ‘GHC.Show’))
+--       mempty :: forall a. Monoid a => a
+--         with mempty @(Int -> String)
+--         (imported from ‘Prelude’ at ...Development/IDE/LSP/CodeAction.hs:7:8-37
+--          (and originally defined in ‘GHC.Base’)) (lsp-ui)
+
+    | topOfHoleFitsMarker `T.isInfixOf` _message = let
+      findSuggestedHoleFits :: T.Text -> [T.Text]
+      findSuggestedHoleFits = extractFitNames . selectLinesWithFits . dropPreceding . T.lines
+      proposeHoleFit name = ("replace hole `" <> holeName <>  "` with " <> name, [TextEdit _range name])
+      holeName = T.strip $ last $ T.splitOn ":" $ head . T.splitOn "::" $ head $ filter ("Found hole" `T.isInfixOf`) $ T.lines _message
+      dropPreceding       = dropWhile (not . (topOfHoleFitsMarker `T.isInfixOf`))
+      selectLinesWithFits = filter ("::" `T.isInfixOf`)
+      extractFitNames     = map (T.strip . head . T.splitOn " :: ")
+      in map proposeHoleFit $ nubOrd $ findSuggestedHoleFits _message
 
 suggestAction _ _ = []
 
+topOfHoleFitsMarker = if versionBranch compilerVersion >= [8,6]
```

Comment:
`versionBranch` isn’t going to work when compiling against `ghc-lib`. I’ve added a `MIN_GHC_API_VERSION` macro in #73  which should always be used for checking the GHC API version.

Reply:
OK, but this requires CPP, and CI rejected my first attempt because it used CPP, which hlint didn't like. Do I have to permit CPP for this module somewhere?

Reply:
We have a whitelist of modules that are allowed to use `CPP` to make people a bit more cautious when using it. This is definitely a valid usecase, so you can just add your module to the whitelist

Reply:
How do I add the test `Main` to this whitelist?

Reply:
I believe that `Main` should do the trick?

Reply:
Just plain, unqualified `Main`? Tried that already: doesn't work ... unless I'm doing something stupid elsewhere, which is perfectly possible.

Reply:
... yes, I'm being stupid ...

Reply:
Hm not sure why this doesn’t work, maybe @ndmitchell has an idea?

Reply:
I'm struggling to get CPP to work in `Main.hs` of the tests:

It works in `CodeAction.hs`, and I've done exactly the same in `Main.hs` AFAICT, that is:

Add 
````
{-# LANGUAGE CPP #-}
#include "ghc-api-version.h"
````
at the top, then 

````#if MIN_GHC_API_VERSION(8,6,0)````


 I did need to add `include-dirs: include` to `test-suite ghcide-tests` in `ghcide.cabal`, for the `#include` to work at all in `Main.hs'.

Now I get:

````
$ cabal v2-test -w ghc-8.6.4
Build profile: -w ghc-8.6.4 -O1
In order, the following will be built (use -v for more details):
 - ghcide-0.0.2 (test:ghcide-tests) (file test/exe/Main.hs changed)
Preprocessing test suite 'ghcide-tests' for ghcide-0.0.2..
Building test suite 'ghcide-tests' for ghcide-0.0.2..

test/exe/Main.hs:507:0: error:
     error: missing binary operator before token "("
     #if MIN_GHC_API_VERSION(8,6,0)
     
    |
507 | #if MIN_GHC_API_VERSION(8,6,0)
    | ^

test/exe/Main.hs:521:0: error:
     error: missing binary operator before token "("
     #if MIN_GHC_API_VERSION(8,6,0)
     
    |
521 | #if MIN_GHC_API_VERSION(8,6,0)
    | ^
`cc' failed in phase `C pre-processor'. (Exit code: 1)
````

What am I missing?

Reply:
Oh I see what’s going on here. We rely on the `MIN_VERSION` pragmas which only exist if you depend on `ghc` which the test suite does not atm. Not quite sure what a nice solution would be but for now, just adding a dependency on `ghc` is not that bad.

Reply:
Thanks, that solves the problem.

Where shall I document the ghc-dependence in the tests? Comment in the cabal file itself?

Reply:
Yeah I think a comment in the cabal file should be sufficient, thank you!